dial_sync.go

```diff
@@ -53,47 +104,96 @@ func (ad *activeDial) wait(ctx context.Context) (*Conn, error) {
 	}
 }
 
+func (ad *activeDial) dialContext(ctx context.Context) context.Context {
+	dialCtx := ad.ctx
+
+	forceDirect, reason := network.GetForceDirectDial(ctx)
+	if forceDirect {
+		dialCtx = network.WithForceDirectDial(dialCtx, reason)
+	}
+
+	simConnect, reason := network.GetSimultaneousConnect(ctx)
+	if simConnect {
+		dialCtx = network.WithSimultaneousConnect(dialCtx, reason)
+	}
+
+	return dialCtx
+}
+
+func (ad *activeDial) dedup(addrs []ma.Multiaddr) (result []ma.Multiaddr) {
+	ad.addrsLk.Lock()
+	defer ad.addrsLk.Unlock()
+
+	for _, a := range addrs {
```

Comment:
What if we never bother to dial one of these addresses because we're canceled? We'll just skip that overall.

Test case:

1. Start dial A.
2. Start dial B.
3. Cancel dial A.
4. Dial B should pick up where dial A left off.

With the current code, dial B will, I believe, fail to dial any of the addresses dial A was trying to dial.

Reply:
no, the dials will still happen -- they use the background context, so there should be no interference.

Reply:
Ah, I see.